[kaka0 Adult] ProductDetailView 제품상세정보 페이지

jotasic opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 제품상세정보 페이지 요청 기능 추가
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- /products/`product_id` 형식으로 `GET` 요청시에 product_id에 해당되는 제품의 상세정보를 리턴하도록 수정

![image](https://user-images.githubusercontent.com/8219812/121886042-e1d6e700-cd4f-11eb-80cd-18a945cd05bb.png)


## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.) 

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

jotasic committed:
Add: 제품 상세페이지 구현

jotasic committed:
Add: 제품상세페이지 결과 값에 별점을 준 유저가 몇명인지의 대한 data도 추가

jotasic committed:
Remove: db_uploader.py 파일 삭제

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,42 @@
+    @login_decorator
+    def get(self, request, product_id):
+        try:
+            product = get_object_or_404(Product, id=product_id)
```
get_object_or_404 method 잘 찾아서 적용하셨네요!!!

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,42 @@
+            product = get_object_or_404(Product, id=product_id)
+            user    = request.user
+
+            sart_point_info = product.review_set.aggregate(
+                    avg=Coalesce(Avg('star_point'), Value(0.0)),
+                    count=Count('id'))
```
```suggestion
            sart_point = product.review_set.aggregate(
                    avg=Coalesce(Avg('star_point'), Value(0.0)),
                    count=Count('id'))
```
- star_point로도 충분히 어떤 의미인지 알 수 있을 것 같습니다.
- avg, count의 정보를 담고 있는 의미로 사용한거면 information으로 줄이지 말고 사용해 주세요.
- Coalesce의 왜 사용 했고, 어떻게 사용하는지 코멘트로 남겨주세요.

jotasic replied:
> Coalesce의 왜 사용 했고, 

해당 상품 리뷰  별점의 평균을 내야 되었습니다.
일반적으로는 Coalesce가 없어도 잘 됩니다. 다만 상품에 리뷰가 하나라도 없는경우 `null`를 리턴합니다.
저는 상품에 대한 리뷰가 1개도 없을시에도 0으로 나와야 된다고 생각하여서 사용하였습니다.

> 어떻게 사용하는지 코멘트로 남겨주세요

Coalesce(a, b, c, d, 0)와 같이 2개 이상의 인자를 받고, 왼쪽부터 차례대로 검사하여 `null`이 아닌 첫번째 값을 리턴합니다.
따라서 위에같은 경우 평점이 `null`이 될 시에 그 다음 값인 0.0을 반환할 수 있도록 사용 할 수 있었습니다.

jotasic replied:
star_point 로 변수명 변경 완료했습니다.

jotasic replied:
삼항연산자로 None일 경우 0을 반환하는 식의 코드가 Coalesce를 사용했을 때 보다, 코스트 측면에서 더 싸다고 구두로 말씀해주셔서 수정 완료했습니다.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,42 @@
+                    product.imageurl_set.values_list(
+                    'url', flat=True).order_by('id')),
+                'like'           : product.user_set.filter(id=user.id).exists(),
+                'cart'           : product.orderlist_set.all().filter(
```
```suggestion
                'cart'           : product.orderlist_set.filter(
```

jotasic replied:
수정완료했습니다.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,42 @@
+                    'url', flat=True).order_by('id')),
+                'like'           : product.user_set.filter(id=user.id).exists(),
+                'cart'           : product.orderlist_set.all().filter(
+                    order__user__id             = user.id,
```
```suggestion
                    order__user             = user,
```

jotasic replied:
수정완료했습니다.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,42 @@
+                'like'           : product.user_set.filter(id=user.id).exists(),
+                'cart'           : product.orderlist_set.all().filter(
+                    order__user__id             = user.id,
+                    order__order_status__status = 'BASKET').exists(),
```
이 부분은 models.py에서 정의한대로 수정해 주세요.

jotasic replied:
네 order_list 컬럼이 order_item로 변경전 내용이여서 병합후에 수정해서 push하겠습니다.

jotasic replied:
수정완료했습니다.

jotasic committed:
Modify: 병합 후 충돌해결

jotasic committed:
Refactor: ProdcutDetailView 코드 리펙토링
1. order model 변경으로 인한 코드 수정
2. 사용이 불필요한 all() 사용 코드 삭제
3. 사용이 불필요한 user_id 를 user로 대체

jotasic committed:
Fix: order_status 조건 잘못 기입으로 인한 수정

jotasic committed:
Fix: 컨벤션 수정

jotasic committed:
Refactor: 유저리뷰 평균점수 구할 시 사용하는 로직 수정
- 기존에 리뷰를 단 사람이 한명도 없을 경우, Coalesce를 사용해서 null의
  대체값을 0.0으로 반환시키도록 하는 방식으로 하였는데, 멘토님이
삼항연산자를 이용해서 하는 것이 코스트가 더 쌀것 같다고 하셔서 이에
맞도록 수정함

Kang-kyunghun commented:
LGTM~!!
